routers/web/auth/oauth.go

```diff
@@ -536,6 +536,16 @@ func AuthorizeOAuth(ctx *context.Context) {
 // GrantApplicationOAuth manages the post request submitted when a user grants access to an application
 func GrantApplicationOAuth(ctx *context.Context) {
 	form := web.GetForm(ctx).(*forms.GrantApplicationForm)
+
+	if !form.Granted {
+		handleAuthorizeError(ctx, AuthorizeError{
+			State:            ctx.Session.Get("state").(string),
+			ErrorDescription: "the request is denied",
+			ErrorCode:        ErrorCodeAccessDenied,
+		}, ctx.Session.Get("redirect_uri").(string))
+		return
+	}
```

Comment:
Ideally this `if` should be put after `if ctx.Session.Get("client_id") != form.ClientID || ctx.Session.Get("state") != form.State || ...`

Then it could guarantee that the `state` and `redirect_uri` are all right, then you could use `form.State` and `form.RedirectURI` directly.